mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp

```diff
@@ -0,0 +1,468 @@
+    sourceOffset = dBuilder.create<AddIOp>(
+        loc, sourceOffset, dBuilder.create<MulIOp>(loc, mnPerMfmaGroup, d_i));
+    sourceOffset = kBuilder.create<AddIOp>(
+        loc, sourceOffset, kBuilder.create<MulIOp>(loc, dPerBlock, k_i));
+  } else {
+    // srcOffset = (k_i * input_span_per_mfma + blk_id) * MN + blk_td + mn_i
+    // * input_span_length;
+    Value inputSpanLenConstantOp =
+        builder.create<ConstantIndexOp>(loc, inputSpanLen);
+    Value inputSpansPerMfmaInConstantOp =
+        builder.create<ConstantIndexOp>(loc, inputSpansPerMfmaIn);
+    Value blk_id = builder.create<DivUIOp>(loc, laneId, inputSpanLenConstantOp);
+    Value blk_td = builder.create<RemUIOp>(loc, laneId, inputSpanLenConstantOp);
+
+    sourceOffset = builder.create<AddIOp>(loc, sourceOffset, blk_td);
+    sourceOffset = dBuilder.create<AddIOp>(
+        loc, sourceOffset,
+        dBuilder.create<MulIOp>(loc, inputSpanLenConstantOp, d_i));
+    sourceOffset = kBuilder.create<AddIOp>(
+        loc, sourceOffset,
+        kBuilder.create<MulIOp>(
+            loc,
+            kBuilder.create<AddIOp>(
+                loc,
+                kBuilder.create<MulIOp>(loc, k_i,
+                                        inputSpansPerMfmaInConstantOp),
+                blk_id),
+            dPerBlock));
+  }
+  return sourceOffset;
+}
+
+// **************************
+// Wmma accelerator interface
+// **************************
+
+WmmaEmitter::WmmaEmitter(WmmaInsn wmmaInsn, StringRef arch,
+                         XdlopsGemmParamsAttr tuningParams)
+    : AccelEmitter{arch, tuningParams,
+                   initAccelEmitterParams(wmmaInsn, tuningParams)},
+      wmmaInsn(wmmaInsn) {}
+
+AccelEmitterParams
+WmmaEmitter::initAccelEmitterParams(WmmaInsn wmmaInsn,
+                                    XdlopsGemmParamsAttr tuningParams) {
+  AccelEmitterParams params;
+
+  // Extract relevant tuning parameters
+  int64_t kPerBlock = tuningParams.getKPerBlock();
+  int64_t kPack = tuningParams.getKpack();
+
+  params.mRepeats = wmmaInsn.mRepeats;
+  params.nRepeats = wmmaInsn.nRepeats;
+  params.nResultVectors = 1;
+  params.kPerThread = kPerBlock;
+  params.kBase = wmmaInsn.inputLen;
+  params.mPerAccel = wmmaInsn.inputLen;
+  params.nPerAccel = wmmaInsn.inputLen;
+  params.kBasePerThread = kPerBlock * kPack / params.kBase;
+
+  params.argTypeA = wmmaInsn.argTypeA;
+  params.argTypeB = wmmaInsn.argTypeB;
+  params.accVectorType = wmmaInsn.retType;
+
+  return params;
+}
+
+Value WmmaEmitter::computeLdsSourceOffset(OpBuilder &kBuilder, Value k_i,
+                                          OpBuilder &dBuilder, Value d_i,
+                                          OpBuilder &builder, Value dPerBlock,
+                                          Location loc, Value sourceBase,
+                                          Value laneId) {
+
+  // srcOffset = k_i * MN + (laneId % wmmaInputLen) + wmmaInputLen * mn_i;
```

Comment:
[Question] Looks very suspicious that this can give us the right result. Could you remind me how this works? It looks to me that if we have 64 lanes and wmmaInputLen = 16. Then:
 - lane0's offset = k_i * MN + 0 + 16 * mn_i
 - lane16's offset = k_i * MN + 0 + 16 * mn_i

Since there are 4 x 16 lanes within each wave, are they just duplicating in loading from the same place 4 times?

Reply:
Yes, this is exactly how `wmma` works (https://gpuopen.com/learn/wmma_on_rdna3/)!